Mathlib/Algebra/Group/Pi.lean

```diff
@@ -0,0 +1,656 @@
+/-
+Copyright (c) 2018 Simon Hudon. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Simon Hudon, Patrick Massot
+
+! This file was ported from Lean 3 source module algebra.group.pi
+! leanprover-community/mathlib commit b3f25363ae62cb169e72cd6b8b1ac97bacf21ca7
+! Please do not edit these lines, except to modify the commit id
+! if you have ported upstream changes.
+-/
+import Mathlib.Logic.Pairwise
+import Mathlib.Algebra.Hom.GroupInstances
+import Mathlib.Data.Pi.Algebra
+import Mathlib.Data.Set.Function
+
+/-!
+# Pi instances for groups and monoids
+
+This file defines instances for group, monoid, semigroup and related structures on Pi types.
+-/
+
+/-
+  Porting notes:
+
+  See this Zulip discussion: [https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/not.20porting.20pi_instance]
+
+  * This file now includes the pi instances for `AddMonoidWithOne` and `AddGroupWithOne`
+  * This file relied on the `pi_instance` tactic, which was not available at the time of porting.
+    The comment `--pi_instance` is inserted before all fields which were previously derived by
+    `pi_instance`.
+  * This file previously gave data fields explicitly. Now previously-defined instances are sourced
+    via `with` as much as possible.
+-/
+
+universe u v w
+
+variable {ι α : Type _}
+
+variable {I : Type u}
+
+-- The indexing type
+variable {f : I → Type v}
+
+-- The family of types already equipped with instances
+variable (x y : ∀ i, f i) (i : I)
+
+@[to_additive]
+theorem Set.preimage_one {α β : Type _} [One β] (s : Set β) [Decidable ((1 : β) ∈ s)] :
+    (1 : α → β) ⁻¹' s = if (1 : β) ∈ s then Set.univ else ∅ :=
+  Set.preimage_const 1 s
+#align set.preimage_one Set.preimage_one
+#align set.preimage_zero Set.preimage_zero
+
+namespace Pi
+
+@[to_additive]
+instance semigroup [∀ i, Semigroup <| f i] : Semigroup (∀ i : I, f i) :=
+  { mul := (· * ·)
+    --pi_instance
+    mul_assoc := by rename_i inst; intros; ext i; exact (inst i).mul_assoc _ _ _ }
```

Comment:
```suggestion
    mul_assoc := by intros; ext; exact mul_assoc _ _ _ }
```
Why bring the instance explicitly into it? Type class inference can find it just fine. Similarly for many many other cases below.

Reply:
Do we want to go through type class inference when we have the instance _right there_? I thought that would be slower and prone to error on the off chance it somehow found the wrong instance. Maybe this all gets compiled away, though, I'm not sure—in which case we indeed wouldn't need the `inst`s.

Reply:
No, it prefers local variables over everything else so there is no chance of finding the wrong instance. We should not ever refer to instances directly unless there is a specific reason we can't do it otherwise. Don't micromanage typeclass inference because it makes the code more brittle if the instances change.

Reply:
Alright, neat, I'll clean up all of the definitions. :)

Reply:
Looks like it built! `rename_i`'s are gone.